### PR TITLE
Make webgl behave better with session history

### LINKED ISSUE #20924: Going back in history from three.js example panics
Visit https://threejs.org/examples/, click and example, then go back in history.
```
ERROR 2018-06-05T23:33:48Z: canvas::webgl_thread: Last GL operation failed: Clear(17664)
assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286) (thread WebGLThread, at components/canvas/webgl_thread.rs:1030)
stack backtrace:
   0:        0x10831414e - backtrace::backtrace::trace::h0217a35a27c6669c
   1:        0x108314c2c - _$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$::default::he8a822a2cccc31d4
   2:        0x108314c7d - backtrace::capture::Backtrace::new::hc727cc4a6b9e93b8
   3:        0x105bcf79e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h619243cbcb02c412
   4:        0x108323dce - std::panicking::rust_panic_with_hook::h90a47ec287d0101d
   5:        0x108323900 - std::panicking::begin_panic_fmt::h7be80a894b2b7fab
   6:        0x107510347 - canvas::webgl_thread::WebGLImpl::apply::h13716cfaea649d87
   7:        0x10750b150 - _$LT$canvas..webgl_thread..WebGLThread$LT$VR$C$$u20$OB$GT$$GT$::handle_webgl_command::h6a534ef73f67943e
   8:        0x1074df220 - std::sys_common::backtrace::__rust_begin_short_backtrace::h990fc7ccc23ba878
   9:        0x1074b5334 - std::panicking::try::do_call::h66683ea52a2df1d6 (.llvm.9678968464668631832)
  10:        0x10834632e - __rust_maybe_catch_panic
  11:        0x1074e6d72 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::he1afdd2e343b2b00
  12:        0x108337067 - std::sys_common::thread::start_thread::hfa2857da5a919ab1
  13:        0x10831d378 - std::sys::unix::thread::Thread::new::thread_start::h8abec15fd1655cac
  14:     0x7fffcec8293a - _pthread_body
  15:     0x7fffcec82886 - _pthread_start
```

### LINKED ISSUE #20930: three.js geometry colors lookuptable example has WebGL error
https://threejs.org/examples/webgl_geometry_colors_lookuptable.html
There were a bunch of `THREE.WebGLAttributes: Unsupported data buffer format: Float64Array.` in the output, and eventually after pressing buttons I got:
```
assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286) (thread WebGLThread, at components/canvas/webgl_thread.rs:1030)
stack backtrace:
   0:        0x102f7214e - backtrace::backtrace::trace::h0217a35a27c6669c
   1:        0x102f72c2c - _$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$::default::he8a822a2cccc31d4
   2:        0x102f72c7d - backtrace::capture::Backtrace::new::hc727cc4a6b9e93b8
   3:        0x10082d79e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h619243cbcb02c412
   4:        0x102f81dce - std::panicking::rust_panic_with_hook::h90a47ec287d0101d
   5:        0x102f81900 - std::panicking::begin_panic_fmt::h7be80a894b2b7fab
   6:        0x10216e347 - canvas::webgl_thread::WebGLImpl::apply::h13716cfaea649d87
   7:        0x102169150 - _$LT$canvas..webgl_thread..WebGLThread$LT$VR$C$$u20$OB$GT$$GT$::handle_webgl_command::h6a534ef73f67943e
   8:        0x10213d220 - std::sys_common::backtrace::__rust_begin_short_backtrace::h990fc7ccc23ba878
   9:        0x102113334 - std::panicking::try::do_call::h66683ea52a2df1d6 (.llvm.9678968464668631832)
  10:        0x102fa432e - __rust_maybe_catch_panic
  11:        0x102144d72 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::he1afdd2e343b2b00
  12:        0x102f95067 - std::sys_common::thread::start_thread::hfa2857da5a919ab1
  13:        0x102f7b378 - std::sys::unix::thread::Thread::new::thread_start::h8abec15fd1655cac
  14:     0x7fffcec8293a - _pthread_body
  15:     0x7fffcec82886 - _pthread_start
```

### LINKED ISSUE #20934: three.js geometry text example has webgl error
https://threejs.org/examples/webgl_geometry_text.html
```
ERROR 2018-06-06T00:11:18Z: canvas::webgl_thread: Last GL operation failed: Clear(17664)
assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286) (thread WebGLThread, at components/canvas/webgl_thread.rs:1030)
stack backtrace:
   0:        0x11119614e - backtrace::backtrace::trace::h0217a35a27c6669c
   1:        0x111196c2c - _$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$::default::he8a822a2cccc31d4
   2:        0x111196c7d - backtrace::capture::Backtrace::new::hc727cc4a6b9e93b8
   3:        0x10ea5179e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h619243cbcb02c412
   4:        0x1111a5dce - std::panicking::rust_panic_with_hook::h90a47ec287d0101d
   5:        0x1111a5900 - std::panicking::begin_panic_fmt::h7be80a894b2b7fab
   6:        0x110392347 - canvas::webgl_thread::WebGLImpl::apply::h13716cfaea649d87
   7:        0x11038d150 - _$LT$canvas..webgl_thread..WebGLThread$LT$VR$C$$u20$OB$GT$$GT$::handle_webgl_command::h6a534ef73f67943e
   8:        0x110361220 - std::sys_common::backtrace::__rust_begin_short_backtrace::h990fc7ccc23ba878
   9:        0x110337334 - std::panicking::try::do_call::h66683ea52a2df1d6 (.llvm.9678968464668631832)
  10:        0x1111c832e - __rust_maybe_catch_panic
  11:        0x110368d72 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::he1afdd2e343b2b00
  12:        0x1111b9067 - std::sys_common::thread::start_thread::hfa2857da5a919ab1
  13:        0x11119f378 - std::sys::unix::thread::Thread::new::thread_start::h8abec15fd1655cac
  14:     0x7fffcec8293a - _pthread_body
  15:     0x7fffcec82886 - _pthread_start
```

### LINKED ISSUE #20950: three.js md2 loader example has webgl error
https://threejs.org/examples/webgl_loader_md2.html

```
ERROR 2018-06-06T00:41:24Z: canvas::webgl_thread: Last GL operation failed: Clear(17664)
assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286) (thread WebGLThread, at components/canvas/webgl_thread.rs:1030)
stack backtrace:
   0:        0x10506714e - backtrace::backtrace::trace::h0217a35a27c6669c
   1:        0x105067c2c - _$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$::default::he8a822a2cccc31d4
   2:        0x105067c7d - backtrace::capture::Backtrace::new::hc727cc4a6b9e93b8
   3:        0x10292279e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h619243cbcb02c412
   4:        0x105076dce - std::panicking::rust_panic_with_hook::h90a47ec287d0101d
   5:        0x105076900 - std::panicking::begin_panic_fmt::h7be80a894b2b7fab
   6:        0x104263347 - canvas::webgl_thread::WebGLImpl::apply::h13716cfaea649d87
   7:        0x10425e150 - _$LT$canvas..webgl_thread..WebGLThread$LT$VR$C$$u20$OB$GT$$GT$::handle_webgl_command::h6a534ef73f67943e
   8:        0x104232220 - std::sys_common::backtrace::__rust_begin_short_backtrace::h990fc7ccc23ba878
   9:        0x104208334 - std::panicking::try::do_call::h66683ea52a2df1d6 (.llvm.9678968464668631832)
  10:        0x10509932e - __rust_maybe_catch_panic
  11:        0x104239d72 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::he1afdd2e343b2b00
  12:        0x10508a067 - std::sys_common::thread::start_thread::hfa2857da5a919ab1
  13:        0x105070378 - std::sys::unix::thread::Thread::new::thread_start::h8abec15fd1655cac
  14:     0x7fffcec8293a - _pthread_body
  15:     0x7fffcec82886 - _pthread_start
```

### LINKED ISSUE #20953: three.js obj2 meshspray loader example has webgl error
* [x] https://threejs.org/examples/webgl_loader_obj2_meshspray.html
* [ ] https://threejs.org/examples/webgl_loader_vtk.html
* [x] https://threejs.org/examples/webgl_buffergeometry_constructed_from_geometry.html

```
ERROR 2018-06-06T00:47:08Z: canvas::webgl_thread: Last GL operation failed: Clear(17664)
assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286) (thread WebGLThread, at components/canvas/webgl_thread.rs:1030)
stack backtrace:
   0:        0x10ab0114e - backtrace::backtrace::trace::h0217a35a27c6669c
   1:        0x10ab01c2c - _$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$::default::he8a822a2cccc31d4
   2:        0x10ab01c7d - backtrace::capture::Backtrace::new::hc727cc4a6b9e93b8
   3:        0x1083bc79e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h619243cbcb02c412
   4:        0x10ab10dce - std::panicking::rust_panic_with_hook::h90a47ec287d0101d
   5:        0x10ab10900 - std::panicking::begin_panic_fmt::h7be80a894b2b7fab
   6:        0x109cfd347 - canvas::webgl_thread::WebGLImpl::apply::h13716cfaea649d87
   7:        0x109cf8150 - _$LT$canvas..webgl_thread..WebGLThread$LT$VR$C$$u20$OB$GT$$GT$::handle_webgl_command::h6a534ef73f67943e
   8:        0x109ccc220 - std::sys_common::backtrace::__rust_begin_short_backtrace::h990fc7ccc23ba878
   9:        0x109ca2334 - std::panicking::try::do_call::h66683ea52a2df1d6 (.llvm.9678968464668631832)
  10:        0x10ab3332e - __rust_maybe_catch_panic
  11:        0x109cd3d72 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::he1afdd2e343b2b00
  12:        0x10ab24067 - std::sys_common::thread::start_thread::hfa2857da5a919ab1
  13:        0x10ab0a378 - std::sys::unix::thread::Thread::new::thread_start::h8abec15fd1655cac
  14:     0x7fffcec8293a - _pthread_body
  15:     0x7fffcec82886 - _pthread_start
```

### LINKED ISSUE #22977: Inactive pages in session history with webgl content keep running
For example, if you visit https://threejs.org/examples/, choose a demo, watch it for a bit, then go backwards in history, the console output shows that the page is still executing.  At minimum we should not be sending triggering rAF callbacks on pages that are not displaying to the user; possibly we should take even more drastic steps in order to conserve resources (such as automatically discarding any page with a webgl canvas when it becomes inactive).

### LINKED ISSUE #22987: Going forwards in history to a webgl page panics
STR:
1. load https://threejs.org/examples
2. choose the cloth example
3. watch it load
4. go back in history
5. go forward in history

Expected:
Cloth example continues running.

Actual:
```
[2019-03-06T21:00:19Z ERROR canvas::webgl_thread] Last GL operation failed: BindBuffer(34962, Some(WebGLBufferId(2)))
[2019-03-06T21:00:19Z WARN  servo] Panic hook called.
assertion failed: `(left == right)`
  left: `1286`,
 right: `0`: Unexpected WebGL error: 0x506 (1286) (thread WebGLThread, at components/canvas/webgl_thread.rs:1437)
stack backtrace:
   0:        0x10fe07c34 - __ZN9backtrace9backtrace5trace17h9f3458ac04360c1aE
                        at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/backtrace/mod.rs:42
   1:        0x10fe025dc - __ZN9backtrace7capture9Backtrace14new_unresolved17h21921a7295046aa9E
                        at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:88
   2:        0x10fe0253e - __ZN9backtrace7capture9Backtrace3new17h18226acc03ed089aE
                        at /Users/jdm/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.9/src/capture.rs:63
   3:        0x109a61065 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h6511ddc8f2676f98E
                        at ports/servo/non_android_main.rs:110
   4:        0x11121cf90 - __ZN3std9panicking20rust_panic_with_hook17hc96a2ab9717cf47bE
                        at src/libstd/panicking.rs:482
   5:        0x11121c9dc - __ZN3std9panicking18continue_panic_fmt17h829e4b1e6e9a8f39E
                        at src/libstd/panicking.rs:385
   6:        0x11121c930 - __ZN3std9panicking15begin_panic_fmt17h3a4ef0ccd095e4a0E
                        at src/libstd/panicking.rs:340
   7:        0x10e5ccda6 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h917b1f0ddac611b4E
                        at components/canvas/webgl_thread.rs:1437
   8:        0x10e60138f - __ZN6canvas10gl_context16GLContextWrapper13apply_command17hf77262daa18fd2eeE
                        at components/canvas/gl_context.rs:164
   9:        0x10e5bbaa1 - __ZN52_$LT$canvas..webgl_thread..WebGLThread$LT$VR$GT$$GT$20handle_webgl_command17hd3d963302b8b7085E
                        at components/canvas/webgl_thread.rs:207
  10:        0x10e5bada9 - __ZN52_$LT$canvas..webgl_thread..WebGLThread$LT$VR$GT$$GT$10handle_msg17h7f91a54c252d5633E
                        at components/canvas/webgl_thread.rs:169
  11:        0x10e5ba76e - __ZN52_$LT$canvas..webgl_thread..WebGLThread$LT$VR$GT$$GT$5start28_$u7b$$u7b$closure$u7d$$u7d$17h84f50790bf709a1eE
                        at components/canvas/webgl_thread.rs:106
  12:        0x10e5e66e3 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h56911288a4a75b0fE
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/sys_common/backtrace.rs:136
  13:        0x10e736d50 - __ZN3std6thread7Builder15spawn_unchecked28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h1648794bce76646aE
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/thread/mod.rs:469
  14:        0x10e5d4a83 - __ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17h29285d785e46d502E
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/panic.rs:309
  15:        0x10e5e6c3f - __ZN3std9panicking3try7do_call17h299d8907f865ed34E
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/panicking.rs:297
  16:        0x11122621e - ___rust_maybe_catch_panic
                        at src/libpanic_unwind/lib.rs:87
  17:        0x10e5e6b24 - __ZN3std9panicking3try17hb778a41606c7c91dE
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/panicking.rs:276
  18:        0x10e5e6983 - __ZN3std5panic12catch_unwind17hb221ebc2569bff03E
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/panic.rs:388
  19:        0x10e736923 - __ZN3std6thread7Builder15spawn_unchecked28_$u7b$$u7b$closure$u7d$$u7d$17h0dc9110979604618E
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/libstd/thread/mod.rs:468
  20:        0x10e745447 - __ZN50_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$8call_box17h5ad430700f108549E
                        at /rustc/0ea22717a1e01fa535534b85a5347a7e49fc79de/src/liballoc/boxed.rs:749
  21:        0x11122564b - __ZN3std3sys4unix6thread6Thread3new12thread_start17hb65474f74178fcb3E
                        at src/libstd/sys_common/thread.rs:14
  22:     0x7fffe4b1793a - __pthread_body
  23:     0x7fffe4b17886 - __pthread_start
```

### PR DESCRIPTION
**Pull request by jdm**
This prevents the compositor from animating pages that are not actually visible, so pages using webgl do not needlessly impact the performance of the rest of the browser. Additionally, this fixes a problem that was alluded to in [this code](https://github.com/servo/rust-offscreen-rendering-context/blob/b5228c098b889a9806a5f93582903e192b3939ef/src/draw_buffer.rs#L282-L285), causing Servo to delete arbitrary resources when a GC occurred in content that used three.js.

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22987 and fix #22977 and fix #20934 and fix #20953 and fix #20930 and fix #20950 and fix #20924
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23000)

### CODE REVIEW

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -2655,6 +2655,13 @@ where
                 },
             };
 
+        if let Some(old_pipeline) = self.pipelines.get(&old_pipeline_id) {
+            old_pipeline.notify_visibility(false);
+        }
+        if let Some(new_pipeline) = self.pipelines.get(&new_pipeline_id) {
+            new_pipeline.notify_visibility(true);
+        }
+
```
Comment (asajeffrey) [inline components/constellation/constellation.rs]: I could have sworn we were already doing this. Am I going mad, and just dreaming up code? @cbrewster, can you remember?

**Review by asajeffrey — APPROVED** [review]
OMG please yes can we have this yesterday?

File: components/constellation/constellation.rs (hunk 1 above)
Comment (jdm) [reply components/constellation/constellation.rs]: See #22986 - the visibility stuff is totally unused at the moment.

Comment (cbrewster) [reply components/constellation/constellation.rs]: It's been a while. I think we only updated activity, but not visibility.

**Comment by jdm** [comment]
The automated test relies on the behaviour changes from #22999, so this PR will need to wait until that one is merged.

**Comment by nox** [comment]
@bors-servo r+

Nice.

**Comment by CYBAI** [comment]
```
{
    "status": "FAIL", 
    "group": "default", 
    "message": "/_mozilla/mozilla/webgl/clearcolor_green.html 481577d73ed0b1f93ad922c251e4e029ec2b8660\n/_mozilla/mozilla/webgl/clearcolor_ref.html b28acfb10eb9c8254d2c7994e1a35f5c0aa70199\nTesting 481577d73ed0b1f93ad922c251e4e029ec2b8660 == b28acfb10eb9c8254d2c7994e1a35f5c0aa70199", 
    "stack": null, 
    "subtest": null, 
    "test": "/_mozilla/mozilla/webgl/clearcolor_green.html", 
    "line": 42656, 
    "action": "test_result", 
    "expected": "PASS"
}
{
    "status": "FAIL", 
    "group": "default", 
    "message": "/_mozilla/mozilla/webgl/clearcolor_blue.html bb883ed4bef46b50c14777ea035a95ea54ee4307\n/_mozilla/mozilla/webgl/clearcolor_ref.html b28acfb10eb9c8254d2c7994e1a35f5c0aa70199\nTesting bb883ed4bef46b50c14777ea035a95ea54ee4307 == b28acfb10eb9c8254d2c7994e1a35f5c0aa70199", 
    "stack": null, 
    "subtest": null, 
    "test": "/_mozilla/mozilla/webgl/clearcolor_blue.html", 
    "line": 42969, 
    "action": "test_result", 
    "expected": "PASS"
}
```

**Pull request merged**